### ISSUE TITLE
RE_CONFIG using wrong sequence number of re-initiating connection

### ISSUE THREAD

**Issue by boivie** [sctplab/usrsctp#567]
Scenario: Two peers, A and B. You can consider A the client and B the server.

A connects to B, and (as the INIT sequence was not captured, these are approximations):
A: Initial TSN ~3212112547  
B: Initial TSN ~2928509502 

They communicate for a while, but then A crashes and restarts, and reconnects and sends INIT, and B sends INIT_ACK:

```
A: INIT chunk (Outbound streams: 1000, inbound streams: 1000)
...
    Initial TSN: 1572596602

B: INIT_ACK chunk (Outbound streams: 1000, inbound streams: 2048)
...
    Initial TSN: 3581572922
```

B is a bit surprised about this and - either because the ULP receives a callback about the connection or because of A sends a DATA message indicating that it has connected - wants to reset its stream ID 4, so it sends a:

```
B: RE_CONFIG chunk
    Chunk type: RE_CONFIG (130)
    Chunk flags: 0x00
    Chunk length: 22
    Outgoing SSN reset request parameter
        Parameter type: Outgoing SSN reset request (0x000d)
        Parameter length: 18
        Re-configuration request sequence number: 2928509497
        Re-configuration response sequence number: 3212112545
        Senders last assigned TSN: 2928509525
        Stream Identifier: 4
    Chunk padding: 0000
```

usrsctp library in A will think those sequence numbers are wrong, as they were from the old association (which it's not aware of), so it will respond:

```
A: RE_CONFIG chunk
    Chunk type: RE_CONFIG (130)
    Chunk flags: 0x00
    Chunk length: 16
    Re-configuration response parameter
        Parameter type: Re-configuration response (0x0010)
        Parameter length: 12
        Re-configuration response sequence number: 2928509497
        Result: Error - Bad sequence number (5)
```

So: The issue here is that the RE_CONFIG uses old sequence number and they should be updated when a new association is made. The RFC6525 isn't quite covering this, but as it is now, B will continuously try to retransmit those RE_CONFIG and the effect isn't as you would expect.

**Comment by tuexen** [comment]
Are both sides using usrsctp?

**Comment by boivie** [comment]
Yes

**Comment by tuexen** [comment]
Let me test this locally...

**Comment by tuexen** [comment]
If I understand your description correct, you are saying that if the peer restarts, and the local node sends out an outgoing stream reset request, it is using the old sequence numbers. Is that correct?

I tested this with the latest kernel stack on FreeBSD by using this packetdrill script:
```
 0.0 `sysctl -w net.inet.sctp.ecn_enable=0`
+0.0 `sysctl -w net.inet.sctp.pr_enable=0`
+0.0 `sysctl -w net.inet.sctp.asconf_enable=0`
+0.0 `sysctl -w net.inet.sctp.auth_enable=0`
+0.0 `sysctl -w net.inet.sctp.reconfig_enable=1`
+0.0 `sysctl -w net.inet.sctp.nrsack_enable=0`
+0.0 `sysctl -w net.inet.sctp.pktdrop_enable=0`
+0.0 socket(..., SOCK_STREAM, IPPROTO_SCTP) = 3
+0.0 setsockopt(3, IPPROTO_SCTP, SCTP_ENABLE_STREAM_RESET, {assoc_id=0,
                                                            assoc_value=SCTP_ENABLE_RESET_STREAM_REQ}, 8) = 0
+0.0 bind(3, ..., ...) = 0
+0.0 listen(3, 1) = 0
+0.0 < sctp: INIT[flgs=0, tag=1, a_rwnd=1500, os=2, is=2, tsn=42,
                  SUPPORTED_EXTENSIONS[types=[RECONFIG]]]
+0.0 > sctp: INIT_ACK[flgs=0, tag=..., a_rwnd=..., os=2, is=2048, tsn=100,
                      SUPPORTED_EXTENSIONS[types=[RECONFIG]],
                      STATE_COOKIE[len=..., val=...]]
+0.0 < sctp: COOKIE_ECHO[flgs=0, len=..., val=...]
+0.0 > sctp: COOKIE_ACK[flgs=0]
+0.0 accept(3, ..., ...) = 4
// Check that we have 2 incoming and outgoing streams.
+0.0 getsockopt(4, IPPROTO_SCTP, SCTP_STATUS, {sstat_assoc_id=...,
                                               sstat_state=SCTP_ESTABLISHED,
                                               sstat_rwnd=...,
                                               sstat_unackdata=...,
                                               sstat_penddata=...,
                                               sstat_instrms=2,
                                               sstat_outstrms=2,
                                               sstat_fragmentation_point=...,
                                               sstat_primary=...}, [176]) = 0
+1.0 < sctp: INIT[flgs=0, tag=2, a_rwnd=1500, os=1, is=1, tsn=84,
                  SUPPORTED_EXTENSIONS[types=[RECONFIG]]]
+0.0 > sctp: INIT_ACK[flgs=0, tag=..., a_rwnd=..., os=1, is=2048, tsn=200,
                      SUPPORTED_EXTENSIONS[types=[RECONFIG]],
                      STATE_COOKIE[len=..., val=...]]
+0.0 < sctp: COOKIE_ECHO[flgs=0, len=..., val=...]
+0.0 > sctp: COOKIE_ACK[flgs=0]
// Check that we have 1 incoming and outgoing stream.
+0.0 getsockopt(4, IPPROTO_SCTP, SCTP_STATUS, {sstat_assoc_id=...,
                                               sstat_state=SCTP_ESTABLISHED,
                                               sstat_rwnd=...,
                                               sstat_unackdata=...,
                                               sstat_penddata=...,
                                               sstat_instrms=1,
                                               sstat_outstrms=1,
                                               sstat_fragmentation_point=...,
                                               sstat_primary=...}, [176]) = 0
+0.0 < sctp: DATA[flgs=BE, len=1016, tsn=84, sid=0, ssn=0, ppid=0]
+0.0 > sctp: SACK[flgs=0, cum_tsn=84, a_rwnd=..., gaps=[], dups=[]]
+0.0 recv(4, ..., 2000, 0) = 1000
+0.0 setsockopt(4, IPPROTO_SCTP, SCTP_RESET_STREAMS, {srs_assoc_id=0,
                                                      srs_flags=SCTP_STREAM_RESET_OUTGOING,
                                                      srs_number_streams=1,
                                                      srs_stream_list=[0]}, 12) = 0
+0.0 > sctp: RECONFIG[flgs=0, OUTGOING_SSN_RESET[req_sn=200, resp_sn=83, last_tsn=199, sids=[]]]
+0.0 close(4) = 0
+0.0 close(3) = 0
```
Here is the resulting tracefile: [issue-567.pcapng.gz](https://github.com/sctplab/usrsctp/files/5883451/issue-567.pcapng.gz)

The behaviour looks correct to me. What am I missing? Could it be that you are not testing with the latest sources?

**Comment by boivie** [comment]
Thanks for looking into this. Your capture looks good. I'm not using the very latest sources, but quite new: 4191ca1784. 

Let me try to reproduce this in a hermetic test. I'll probably have it done today, though.

**Comment by tuexen** [comment]
That sounds strange. I don't see any relevant change between the version you mention and the latest code. I also don't expect a difference between the kernel stack and the userland stack with respect to the behaviour you are focussing on. Maybe we are missing something. Let me know when you have a small reproducer. Having a .pcap file would allow me to reproduce the on-wire behaviour with packetdrill.

**Comment by boivie** [comment]
I have tried to reproduce it, but I can't. I do however have multiple captures of this, which I could e-mail you separately.

The issue, I believe, is that the sequence is as follows:
```
	12:06:19.859387	203.0.113.1	127.0.0.1	126	INIT 
	12:06:19.888119	127.0.0.1	203.0.113.1	328	INIT_ACK 
	12:06:19.888250	203.0.113.1	127.0.0.1	664	COOKIE_ECHO DATA 
	12:06:19.891898	127.0.0.1	203.0.113.1	78	DATA_CHANNEL_OPEN_REQUEST 
	12:06:19.891922	127.0.0.1	203.0.113.1	58	RE_CONFIG 
	12:06:19.916953	127.0.0.1	203.0.113.1	56	COOKIE_ACK SACK 
```
So the stream resetting is happening just between COOKIE_ECHO (which creates the new TCB on reconnect) and it sending the COOKIE_ACK, so maybe the Stream Reset used parameters from the previous TCB? 

The stream reset can have been initiated by peer A noticing that peer B "had issues" by other means than SCTP notifications, because I tried to send a outgoing stream reset during processing SCTP_ASSOC_CHANGE with state SCTP_RESTART, but that didn't cause a race.
